source/common/common/safe_memcpy.h

```diff
@@ -0,0 +1,29 @@
+#pragma once
+
+#include <cstring>
+
+namespace Envoy {
+
+/**
+ * @brief Copies src to dst based on their sizes, which must be the same.
+ */
+template <typename T1, typename T2> inline void safeMemcpy(T1* dst, T2* src) {
+  static_assert(sizeof(T1) == sizeof(T2));
+  memcpy(dst, src, sizeof(T1));
+}
+
+/**
+ * @brief Copies src to dst based on the size of dst
+ * @note  Sizes are not compared, so ensure the src is of size sizeof(*(dst)) before proceeding to
+ * call safeMemcpyUnsafeSrc
+ */
+#define safeMemcpyUnsafeSrc(dst, src) memcpy(dst, src, sizeof(*(dst)));
```

Comment:
Any reason those are macro but not template inline?

Reply:
Nothing particularly relevant for this case 😅, although, I guess, it could be argued that with the macro you gurantee code subsitution (whereas with inlining that might not happen) and compatibility with older compilers (previous to c++14). If you provide better reasons to replace this, I will submit safeMemcpyUnsafeSrc and safeMemcpyUnsafeDst as inline function templates on the next commit 😉

Reply:
+1 for templates here. I don't know of a scenario where inlining would not occur for this case, TBH, even prior to C++14.

You can templatize for for this one on the type of `src` and have dst be `void*` -- I think that will inline without issue.